Support bouncyCastle libs  + Infinispan + FIPS

### Describe the bug

Infinispan: 12.1 / 14
Quarkus: 2.7.5.Final
Java 11 / 17

Seems that bouncy castle lib is not supported until Infinispan 14.0. But If I move on to infinispan 14 and configure Quarkus in order to use `BCFIPSJSSE` then I got the following error:

```
13:51:46,057 INFO  [app] Caused by: java.security.KeyStoreException: BCFKS not found
13:51:46,057 INFO  [app]        at java.base/java.security.KeyStore.getInstance(KeyStore.java:871)
13:51:46,057 INFO  [app]        at org.infinispan.commons.util.SslContextFactory.getTrustManagerFactory(SslContextFactory.java:172)
13:51:46,057 INFO  [app]        at org.infinispan.client.hotrod.impl.transport.netty.ChannelInitializer.initSsl(ChannelInitializer.java:154)
13:51:46,057 INFO  [app]        ... 20 more
13:51:46,057 INFO  [app] Caused by: java.security.NoSuchAlgorithmException: BCFKS KeyStore not available
13:51:46,057 INFO  [app]        at java.base/sun.security.jca.GetInstance.getInstance(GetInstance.java:159)
13:51:46,057 INFO  [app]        at java.base/java.security.Security.getImpl(Security.java:701)
13:51:46,057 INFO  [app]        at java.base/java.security.KeyStore.getInstance(KeyStore.java:868)

```


My infisnispan configuration looks like the following one:
```
# Infinispan
quarkus.infinispan-client.trust-store=src/main/resources/server.jks
quarkus.infinispan-client.trust-store-password=password
quarkus.infinispan-client.trust-store-type=BCFKS

quarkus.infinispan-client.server-list=localhost:11222
quarkus.infinispan-client.use-auth=true
```

In my opinion, I am missing some configurations that we already have in orders [modules](https://quarkus.io/guides/security-customization#bouncy-castle-jsse-fips), something like `quarkus.infinispan-client.trust-store-provider=BCFIPS`

[Reproducer](https://github.com/pjgg/quarkus-test-suite/blob/feat/infisnipan_fips_reproducer/messaging/infinispan-grpc-kafka/src/test/java/io/quarkus/qe/messaging/infinispan/InfinispanKafkaIT.java)
```
git clone git@github.com:quarkus-qe/quarkus-test-suite.git
git checkout feat/infisnipan_fips_reproducer
```

Scenario
```
mvn clean verify -Dall-modules -pl messaging/infinispan-grpc-kafka -Dit.test=InfinispanKafkaIT
```

I am not sure if `BCFIPS` is supported and if does then, I am not sure how to use it. Should I add `quarkus-security` extension ?, and If I am able to set `BCFIPS` as a provider I understand that then later I will not need any [JVM security custom](https://docs.oracle.com/middleware/11119/wls/SECMG/fips.htm#A1233313) configuration, isn't it? (on `quarkus.http.ssl` is not need it)

--- a/docs/src/main/asciidoc/infinispan-client.adoc
+++ b/docs/src/main/asciidoc/infinispan-client.adoc
@@ -61,14 +61,14 @@ If you modify `application.properties` or `hotrod-client.properties`, you must r
 
 == Connecting to Infinispan clusters
 
-Add the following properties to connect to Infinispan Server:
+If you are running an Infinispan Server add the following properties to connect:
 
 [source,properties]
 ----
-quarkus.infinispan-client.server-list=localhost:11222 <1>
+quarkus.infinispan-client.hosts=localhost:11222 <1>
 
-quarkus.infinispan-client.auth-username=admin <2>
-quarkus.infinispan-client.auth-password=password <3>
+quarkus.infinispan-client.username=admin <2>
+quarkus.infinispan-client.password=password <3>
 
 quarkus.infinispan-client.client-intelligence=BASIC <4>
 ----
@@ -76,6 +76,20 @@ quarkus.infinispan-client.client-intelligence=BASIC <4>
 <2> Sets the authentication username
 <3> Sets the authentication password
 <4> Sets the client intelligence. Use BASIC as a workaround if using Docker for Mac.
+
+Alternatively, you can use uri connection by providing a single connection property
+[source,properties]
+----
+quarkus.infinispan-client.uri=hotrod://admin:password@localhost:11222 <1>
+quarkus.infinispan-client.client-intelligence=BASIC <2>
+----
+<1> Sets Infinispan URI connection. The following properties will be ignored: hosts, username and password.
+<2> Sets the client intelligence. Use BASIC as a workaround if using Docker for Mac
+
+[TIP]
+====
+Use Infinispan Dev Services to run a server and connect without configuration.
+====
 
 .Running Infinispan Server
 
@@ -92,6 +106,14 @@ Infinispan Server also enables authentication and security authorization by defa
 ----
 $ ./bin/cli.sh user create admin -p password
 ----
+
+=== Infinispan Health Check
+If you are using the quarkus-smallrye-health extension, the Infinispan client extensions will automatically add a readiness health check to validate the connection.
+
+When you access the `/q/health/ready` endpoint of your application you will have information about the server connection and available caches.
+
+This behavior can be disabled via the property `quarkus.infinispan-client.health.enabled`.
+
 
 === Creating caches from the client
 
@@ -483,8 +505,10 @@ You need to register the generated Protobuf schemas with Infinispan Server to pe
 `Protobuf` to other media types such as `JSON`.
 
 [TIP]
+====
 You can check the schemas that exist under the `Schemas` tab by logging into
 Infinispan Console at `http://localhost:11222`
+====
 
 By default, Protobuf schemas generated this way will be registered by this extension when the client first connects.
 However, it might be required to handle the registration manually as a schema may evolve over time when used in
@@ -691,11 +715,12 @@ When you use the infinispan-client extension in dev mode or in test, Quarkus aut
 Dev Services for Infinispan is automatically enabled unless:
 
 - `quarkus.infinispan-client.devservices.enabled` is set to `false`
-- the `quarkus.infinispan-client.server-list` is configured
+- the `quarkus.infinispan-client.hosts` is configured
+- the `quarkus.infinispan-client.uri` is configured
 
 Dev Services for Infinispan relies on Docker to start the broker.
 If your environment does not support Docker, you will need to start the broker manually, or connect to an already running broker.
-You can configure the broker address using `quarkus.infinispan-client.server-list`.
+You can configure the broker address using `quarkus.infinispan-client.hosts`.
 
 == Shared server
 

--- a/docs/src/main/asciidoc/kogito.adoc
+++ b/docs/src/main/asciidoc/kogito.adoc
@@ -407,7 +407,7 @@ Add following into the src/main/resources/application.properties file (create th
 
 [source,plain]
 ----
-quarkus.infinispan-client.server-list=localhost:11222
+quarkus.infinispan-client.hosts=localhost:11222
 ----
 
 NOTE: Adjust the host and port number according to your Infinispan server installation.

--- a/extensions/infinispan-client/deployment/src/main/java/io/quarkus/infinispan/client/deployment/devservices/InfinispanDevServiceProcessor.java
+++ b/extensions/infinispan-client/deployment/src/main/java/io/quarkus/infinispan/client/deployment/devservices/InfinispanDevServiceProcessor.java
@@ -103,7 +103,7 @@ public class InfinispanDevServiceProcessor {
             }
             newDevServices.add(devService);
             log.infof("The infinispan server is ready to accept connections on %s",
-                    devService.getConfig().get(getConfigPrefix() + "server-list"));
+                    devService.getConfig().get(getConfigPrefix() + "hosts"));
             compressor.close();
         } catch (Throwable t) {
             compressor.closeAndDumpCaptured();
@@ -144,14 +144,17 @@ public class InfinispanDevServiceProcessor {
 
         String configPrefix = getConfigPrefix();
 
-        boolean needToStart = !ConfigUtils.isPropertyPresent(configPrefix + "server-list");
+        boolean needToStart = !ConfigUtils.isPropertyPresent(configPrefix + "hosts")
+                && !ConfigUtils.isPropertyPresent(configPrefix + "server-list");
+
         if (!needToStart) {
-            log.debug("Not starting devservices for Infinispan as 'server-list' have been provided");
+            log.debug("Not starting devservices for Infinispan as 'hosts', 'uri' or 'server-list' have been provided");
             return null;
         }
 
         if (!dockerStatusBuildItem.isDockerAvailable()) {
-            log.warn("Please configure 'quarkus.infinispan-client.server-list' or get a working docker instance");
+            log.warn(
+                    "Please configure 'quarkus.infinispan-client.hosts' or 'quarkus.infinispan-client.uri' or get a working docker instance");
             return null;
         }
 
@@ -176,10 +179,10 @@ public class InfinispanDevServiceProcessor {
     private RunningDevService getRunningDevService(String containerId, Closeable closeable, String serverList,
             String username, String password) {
         Map<String, String> config = new HashMap<>();
-        config.put(getConfigPrefix() + "server-list", serverList);
+        config.put(getConfigPrefix() + "hosts", serverList);
         config.put(getConfigPrefix() + "client-intelligence", "BASIC");
-        config.put(getConfigPrefix() + "auth-username", username);
-        config.put(getConfigPrefix() + "auth-password", password);
+        config.put(getConfigPrefix() + "username", username);
+        config.put(getConfigPrefix() + "password", password);
         return new RunningDevService(Feature.INFINISPAN_CLIENT.getName(), containerId, closeable, config);
     }
 

--- a/extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/InfinispanClientProducer.java
+++ b/extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/InfinispanClientProducer.java
@@ -180,8 +180,38 @@ public class InfinispanClientProducer {
         }
         InfinispanClientRuntimeConfig infinispanClientRuntimeConfig = this.infinispanClientRuntimeConfig.get();
 
-        if (infinispanClientRuntimeConfig.serverList.isPresent()) {
-            properties.put(ConfigurationProperties.SERVER_LIST, infinispanClientRuntimeConfig.serverList.get());
+        if (infinispanClientRuntimeConfig.uri.isPresent()) {
+            properties.put(ConfigurationProperties.URI, infinispanClientRuntimeConfig.uri.get());
+        } else {
+            if (infinispanClientRuntimeConfig.serverList.isPresent()) {
+                log.warn(
+                        "Use 'quarkus.infinispan-client.hosts' instead of the deprecated 'quarkus.infinispan-client.server-list'");
+                properties.put(ConfigurationProperties.SERVER_LIST, infinispanClientRuntimeConfig.serverList.get());
+            }
+
+            if (infinispanClientRuntimeConfig.hosts.isPresent()) {
+                properties.put(ConfigurationProperties.SERVER_LIST, infinispanClientRuntimeConfig.hosts.get());
+            }
+
+            if (infinispanClientRuntimeConfig.authUsername.isPresent()) {
+                log.warn(
+                        "Use 'quarkus.infinispan-client.username' instead of the deprecated 'quarkus.infinispan-client.auth-username'");
+                properties.put(ConfigurationProperties.AUTH_USERNAME, infinispanClientRuntimeConfig.authUsername.get());
+            }
+
+            if (infinispanClientRuntimeConfig.username.isPresent()) {
+                properties.put(ConfigurationProperties.AUTH_USERNAME, infinispanClientRuntimeConfig.username.get());
+            }
+
+            if (infinispanClientRuntimeConfig.authPassword.isPresent()) {
+                log.warn(
+                        "Use 'quarkus.infinispan-client.password' instead of the deprecated 'quarkus.infinispan-client.auth-password'");
+                properties.put(ConfigurationProperties.AUTH_PASSWORD, infinispanClientRuntimeConfig.authPassword.get());
+            }
+
+            if (infinispanClientRuntimeConfig.password.isPresent()) {
+                properties.put(ConfigurationProperties.AUTH_PASSWORD, infinispanClientRuntimeConfig.password.get());
+            }
         }
 
         if (infinispanClientRuntimeConfig.clientIntelligence.isPresent()) {
@@ -191,21 +221,19 @@ public class InfinispanClientProducer {
         if (infinispanClientRuntimeConfig.useAuth.isPresent()) {
             properties.put(ConfigurationProperties.USE_AUTH, infinispanClientRuntimeConfig.useAuth.get());
         }
-        if (infinispanClientRuntimeConfig.authUsername.isPresent()) {
-            properties.put(ConfigurationProperties.AUTH_USERNAME, infinispanClientRuntimeConfig.authUsername.get());
-        }
-        if (infinispanClientRuntimeConfig.authPassword.isPresent()) {
-            properties.put(ConfigurationProperties.AUTH_PASSWORD, infinispanClientRuntimeConfig.authPassword.get());
-        }
+
         if (infinispanClientRuntimeConfig.authRealm.isPresent()) {
             properties.put(ConfigurationProperties.AUTH_REALM, infinispanClientRuntimeConfig.authRealm.get());
         }
+
         if (infinispanClientRuntimeConfig.authServerName.isPresent()) {
             properties.put(ConfigurationProperties.AUTH_SERVER_NAME, infinispanClientRuntimeConfig.authServerName.get());
         }
+
         if (infinispanClientRuntimeConfig.authClientSubject.isPresent()) {
             properties.put(ConfigurationProperties.AUTH_CLIENT_SUBJECT, infinispanClientRuntimeConfig.authClientSubject.get());
         }
+
         if (infinispanClientRuntimeConfig.authCallbackHandler.isPresent()) {
             properties.put(ConfigurationProperties.AUTH_CALLBACK_HANDLER,
                     infinispanClientRuntimeConfig.authCallbackHandler.get());
@@ -224,6 +252,18 @@ public class InfinispanClientProducer {
         }
         if (infinispanClientRuntimeConfig.trustStoreType.isPresent()) {
             properties.put(ConfigurationProperties.TRUST_STORE_TYPE, infinispanClientRuntimeConfig.trustStoreType.get());
+        }
+
+        if (infinispanClientRuntimeConfig.sslProvider.isPresent()) {
+            properties.put(ConfigurationProperties.SSL_PROVIDER, infinispanClientRuntimeConfig.sslProvider.get());
+        }
+
+        if (infinispanClientRuntimeConfig.sslProtocol.isPresent()) {
+            properties.put(ConfigurationProperties.SSL_PROTOCOL, infinispanClientRuntimeConfig.sslProtocol.get());
+        }
+
+        if (infinispanClientRuntimeConfig.sslCiphers.isPresent()) {
+            properties.put(ConfigurationProperties.SSL_CIPHERS, infinispanClientRuntimeConfig.sslCiphers.get().toArray());
         }
 
         builder.withProperties(properties);

--- a/extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/InfinispanClientRuntimeConfig.java
+++ b/extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/InfinispanClientRuntimeConfig.java
@@ -1,6 +1,7 @@
 package io.quarkus.infinispan.client.runtime;
 
 import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
 import java.util.Optional;
 
@@ -21,10 +22,31 @@ import io.quarkus.runtime.annotations.ConfigRoot;
 @ConfigRoot(name = "infinispan-client", phase = ConfigPhase.RUN_TIME)
 public class InfinispanClientRuntimeConfig {
 
+    // @formatter:off
+    /**
+     * Sets the URI of the running Infinispan server to connect to. hotrod://localhost:11222@admin:password
+     * If provided {@link #hosts}, {@link #username} and {@link #password} will be ignored.
+     */
+    // @formatter:on
+    @ConfigItem
+    public Optional<String> uri;
+
+    // @formatter:off
     /**
      * Sets the host name/port to connect to. Each one is separated by a semicolon (eg. host1:11222;host2:11222).
      */
+    // @formatter:on
     @ConfigItem
+    public Optional<String> hosts;
+
+    // @formatter:off
+    /**
+     * Sets the host name/port to connect to. Each one is separated by a semicolon (eg. host1:11222;host2:11222).
+     * @deprecated {@link #hosts} should be used to configure the list or uri for an uri connection string.
+     */
+    // @formatter:on
+    @ConfigItem
+    @Deprecated
     public Optional<String> serverList;
 
     // @formatter:off
@@ -66,12 +88,30 @@ public class InfinispanClientRuntimeConfig {
      * Sets username used by authentication.
      */
     @ConfigItem
+    Optional<String> username;
+
+    /**
+     * Sets username used by authentication.
+     *
+     * @deprecated {@link #username} should be used to configure the credentials username.
+     */
+    @ConfigItem
+    @Deprecated
     Optional<String> authUsername;
 
     /**
-     * Sets password used by authentication
+     * Sets password used by authentication.
      */
     @ConfigItem
+    Optional<String> password;
+
+    /**
+     * Sets password used by authentication
+     *
+     * @deprecated {@link #password} should be used to configure the credentials password.
+     */
+    @ConfigItem
+    @Deprecated
     Optional<String> authPassword;
 
     /**
@@ -132,6 +172,27 @@ public class InfinispanClientRuntimeConfig {
      */
     @ConfigItem
     Optional<String> trustStoreType;
+
+    /**
+     * Configures the secure socket protocol.
+     * Setting this property implicitly enables SSL/TLS.
+     */
+    @ConfigItem
+    Optional<String> sslProtocol;
+
+    /**
+     * Sets the ssl provider. For example BCFIPS
+     * Setting this implicitly enables SSL/TLS.
+     */
+    @ConfigItem
+    Optional<String> sslProvider;
+
+    /**
+     * Configures the ciphers.
+     * Setting this property implicitly enables SSL/TLS.
+     */
+    @ConfigItem
+    Optional<List<String>> sslCiphers;
 
     /**
      * Configures caches from the client with the provided configuration.

--- a/test-framework/infinispan-client/src/main/java/io/quarkus/test/infinispan/client/InfinispanTestResource.java
+++ b/test-framework/infinispan-client/src/main/java/io/quarkus/test/infinispan/client/InfinispanTestResource.java
@@ -48,7 +48,7 @@ public class InfinispanTestResource implements QuarkusTestResourceLifecycleManag
         INFINISPAN.start();
 
         final String hosts = INFINISPAN.getHost() + ":" + INFINISPAN.getMappedPort(HOTROD_PORT);
-        return Collections.singletonMap("quarkus.infinispan-client.server-list", hosts);
+        return Collections.singletonMap("quarkus.infinispan-client.hosts", hosts);
     }
 
     @Override

